### PR TITLE
First PR

### PR DESCRIPTION
**Pull request by incogni23**
Updated service.go, consumer.go

### CODE REVIEW

File: .vscode/launch.json [hunk 1]
```diff
@@ -0,0 +1,7 @@
+{
+    // Use IntelliSense to learn about possible attributes.
+    // Hover to view descriptions of existing attributes.
+    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
+    "version": "0.2.0",
+    "configurations": []
+}
```
Comment (mishtuu) [inline .vscode/launch.json]: remove this - P2

use .gitignore

File: .vscode/settings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+{
+    "cmake.configureOnOpen": true
+}
```
Comment (mishtuu) [inline .vscode/settings.json:3]: use .gitignore - P2

File: Timer/clock.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+		NextRetry:          event.NextRetry,
+		Attempts:           event.Attempts,
+	}
+}
+
+func RetryClock() {
+	for {
+		events, err := database.GetIncompleteEvents()
+		if err != nil {
+			fmt.Print("Error getting incomplete events", err)
```
Comment (mishtuu) [inline Timer/clock.go]: use log.Error, log.Info instead of fmt.Printl - P2

and this holds true across all the logs

File: Timer/clock.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+			continue
+		}
+
+		for _, event := range events {
+			if event.NextRetry > time.Now().Unix() {
+				continue
+			}
```
Comment (mishtuu) [inline Timer/clock.go]: is this really needed? - P0


check if you can handle it directly on query level

Comment (mishtuu) [reply Timer/clock.go]: Once you do this, the function `GetIncompleteEvents` name should be changed to something which ties to its name

File: Timer/clock.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+			err := service.Deliver(eventToConsumerEvent(event), event.NotificationType)
+
+			if err != nil {
+				fmt.Print("Error delivering events", err)
+				event.Status = "Not Completed Yet"
```
Comment (mishtuu) [inline Timer/clock.go]: should be enum - P1

Comment (mishtuu) [reply Timer/clock.go]: create a models module which will hold commonly used structs and enums - P0

File: Timer/clock.go [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+				event.Attempts++
+				event.NextRetry = time.Now().Unix() + calculateRetryTime(event.Attempts)
+				err := database.UpdateEvent(&event)
+				if err != nil {
+					fmt.Print("Error in updating event", err)
+				}
+			} else {
+				event.Status = "Completed"
```
Comment (mishtuu) [inline Timer/clock.go]: this COMPLETED string should be the part enum

File: Timer/clock.go [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+		fmt.Print("retry is working")
+		time.Sleep(time.Minute)
+
+	}
+}
+
+func calculateRetryTime(attempts int) int64 {
+	switch attempts {
+	case 0:
+		return 0
```
Comment (mishtuu) [inline Timer/clock.go]: this time should be greater than 0 - P0

File: consumer/consumer.go [hunk 8]
```diff
@@ -1,40 +1,43 @@
...
+	Message            []byte         `json:"message"`
+	Source             string         `json:"source"`
+	DestinationAddress string         `json:"destinationAddress"`
+	Topics             pq.StringArray `gorm:"type:text[]" json:"topics"`
+	NotificationType   string         `json:"notificationtype"`
+	IdempotencyKey     string         `json:"idempotencykey"`
+	Status             string         `json:"status"`
+	Attempts           int            `json:"attempts"`
+	NextRetry          int64          `json:"nextRetry"`
+}
```
Comment (mishtuu) [inline consumer/consumer.go]: should be residing in 1 common module, which is models wala module - P0

File: consumer/consumer.go [hunk 9]
```diff
@@ -1,40 +1,43 @@
...
 	r := kafka.NewReader(kafka.ReaderConfig{
-		Brokers:   brokers,
-		Topic:     Topic,
-		Partition: 0,
+		GroupID: "consumergc",
+		Brokers: brokers,
+		Topic:   Topic,
 	})
-	r.SetOffset(6)
+	//r.SetOffset()
```
Comment (mishtuu) [inline consumer/consumer.go]: remove - P2

File: database/database.go [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import (
+	"github.com/lib/pq"
+	"gorm.io/driver/postgres"
+	"gorm.io/gorm"
+)
+
+var db *gorm.DB
+
+func SetupDB() (*gorm.DB, error) {
+	dsn := "host=localhost user=pikapika password=Ankita@2307 dbname=db1 port=5432 sslmode=disable TimeZone=Asia/Kolkata"
```
Comment (mishtuu) [inline database/database.go]: make it configurable - P2

File: database/database.go [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+	Message            []byte         `json:"message"`
+	Source             string         `json:"source"`
+	DestinationAddress string         `json:"destinationAddress"`
+	Topics             pq.StringArray `gorm:"type:text[]" json:"topics"`
+	NotificationType   string         `json:"notificationtype"`
+	IdempotencyKey     string         `json:"idempotencykey"`
+	Status             string         `json:"status"`
+	Attempts           int            `json:"attempts"`
+	NextRetry          int64          `json:"nextRetry"`
+}
```
Comment (mishtuu) [inline database/database.go]: shift to a common place - P0

File: database/database.go [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+	return db.Create(&event).Error
+}
+
+func UpdateEvent(event *Event) error {
+	return db.Model(&event).Update("Status", event.Status).Error
+}
+
+func GetIncompleteEvents() ([]Event, error) {
+	var events []Event
+	err := db.Where("status = ?", "Not Completed Yet").Find(&events).Error
```
Comment (mishtuu) [inline database/database.go]: Not Completed Yet - should be served from enum

File: producer/producer.go [hunk 13]
```diff
@@ -4,9 +4,24 @@ import (
...
+	Source             string         `json:"source"`
+	DestinationAddress string         `json:"destinationAddress"`
+	Topics             pq.StringArray `gorm:"type:text[]" json:"topics"`
+	NotificationType   string         `json:"notificationtype"`
+	IdempotencyKey     string         `json:"idempotencykey"`
+	Status             string         `json:"status"`
+	Attempts           int            `json:"attempts"`
+	NextRetry          int64          `json:"nextretry"`
+}
+
```
Comment (mishtuu) [inline producer/producer.go]: move it to common place

File: services/notification_services/service.go [hunk 14]
```diff
@@ -0,0 +1,51 @@
...
+	for {
+		consumedEvent := consumerRecieved.Consume()
+		err := Deliver(consumedEvent, consumedEvent.NotificationType)
+		if err != nil {
+			fmt.Print("err in producing msg", err)
+			return
+		}
+	}
+
+}
```
Comment (mishtuu) [inline services/notification_services/service.go]: remove this - P0

File: .vscode/launch.json (hunk 1 above)
Comment (incogni23) [reply .vscode/launch.json]: huh bleh.

File: services/notification_services/service.go [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+import (
+	"bytes"
+	"consumer"
+	"errors"
+	"net/http"
+
+	"github.com/labstack/gommon/log"
+)
+
+func Deliver(event consumer.Event, notificationType string) error {
```
Comment (mishtuu) [inline services/notification_services/service.go]: apart from error, send one more parameter `isRetryble` (bool)

File: services/notification_services/service.go [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+		resp, err := client.Do(req)
+		if err != nil {
+			log.Error("error sending msg to webhook", err)
+			return err
+		}
+		defer resp.Body.Close()
+		log.Info("msg sent", resp.Status)
+
+	default:
+		return errors.New("notification type required")
```
Comment (mishtuu) [inline services/notification_services/service.go]: here we will set isRetryable as false

**Comment by mishtuu** [comment]
max retry attempts should be limited to 5, and this number needs to be configurable

**Comment by mishtuu** [comment]
to make things configurable you can checkout this repo
https://github.com/ilyakaznacheev/cleanenv

**Review by mishtuu — CHANGES_REQUESTED** [review]

File: models/models.go [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+
+const (
+	StatusNotCompleted = "Not Completed"
+	StatusCompleted    = "Completed"
+	StatusIgnored      = "Skipped"
+)
+
+var MaxRetryAttempts = 5
+
+const Topicname string = "quickstart-events" // for consumer
```
Comment (mishtuu) [inline models/models.go]: use env

File: services/notification_services/webhook/service.go [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+			isRetryable = true
+			return err, true
+		}
+		defer resp.Body.Close()
+		log.Info("msg sent", resp.Status)
+		if resp.StatusCode >= 200 && resp.StatusCode <= 299 {
+			isRetryable = true
+		} else {
+			isRetryable = false
+		}
```
Comment (mishtuu) [inline services/notification_services/webhook/service.go]: think and change

Comment (incogni23) [reply services/notification_services/webhook/service.go]: i am not so sure on this one

**Pull request merged**